sd-2019-stack-overflow-frontend/src/view/SmartNavbar.js

```diff
@@ -0,0 +1,80 @@
+import React, {Component} from "react";
+import sOUserModel from "../model/sOUserModel";
+import navbarPresenter from "../presenter/navbarPresenter";
+import NavbarNotLoggedIn from "./NavbarNotLoggedIn";
+import NavbarLoggedIn from "./NavbarLoggedIn";
+import SignUpModal from "./SignUpModal";
+import LogInModal from "./LogInModal";
+
+const mapModelStateToComponentState = (sOUserModelState) => (
+    {
+        loggedInUsername: sOUserModelState.loggedInUsername,
+        signUpModalActive: sOUserModelState.signUpModalActive,
+        newSOUserUsername: sOUserModelState.newSOUser.username,
+        newSOUserPassword: sOUserModelState.newSOUser.password,
+        logInModalActive: sOUserModelState.logInModalActive,
+        logInSOUserUsername: sOUserModelState.logInSOUser.username,
+        logInSOUserPassword: sOUserModelState.logInSOUser.password
+    }
+);
+
+export default class SmartNavbar extends Component {
+    constructor() {
+        super();
+        this.state = mapModelStateToComponentState(sOUserModel.state);
+        this.listener = () => this.setState(mapModelStateToComponentState(sOUserModel.state));
+        sOUserModel.addListener("SOUserModelChange", this.listener);
+    }
+
+    componentWillUnmount() {
+        sOUserModel.removeListener("SOUserModelChange", this.listener);
+    }
+
+    render() {
+        return (
+            <div>
```

Comment:
There is still a little too much presentation going on in this smart component (which would belong in a dumb component normally).